pkgs/development/tools/godot/3/default.nix

```
@@ -1,109 +1,165 @@
-{ stdenv
-, lib
+{ lib
+, stdenv
+, alsa-lib
+, alsa-plugins
+, autoPatchelfHook
 , fetchFromGitHub
-, scons
-, pkg-config
-, udev
+, freetype
+, installShellFiles
+, libGLU
+, libpulseaudio
 , libX11
 , libXcursor
+, libXext
+, libXfixes
+, libXi
 , libXinerama
 , libXrandr
 , libXrender
-, libpulseaudio
-, libXi
-, libXext
-, libXfixes
-, freetype
-, openssl
-, alsa-lib
-, alsa-plugins
 , makeWrapper
-, libGLU
-, zlib
+, openssl
+, pkg-config
+, scons
+, udev
 , yasm
-, withUdev ? true
+, zlib
 }:
 
-let
-  options = {
-    touch = libXi != null;
-    pulseaudio = false;
-    udev = withUdev;
-  };
-in
-stdenv.mkDerivation rec {
-  pname = "godot";
+stdenv.mkDerivation (self: {
+  pname = "godot3";
   version = "3.5.2";
+  buildDescription = "X11 tools";
 
   src = fetchFromGitHub {
     owner = "godotengine";
     repo = "godot";
-    rev = "${version}-stable";
+    rev = "${self.version}-stable";
     sha256 = "sha256-C+1J5N0ETL1qKust+2xP9uB4x9NwrMqIm8aFAivVYQw=";
   };
 
-  nativeBuildInputs = [ pkg-config makeWrapper ];
-  buildInputs = [
+  nativeBuildInputs = [
+    autoPatchelfHook
+    installShellFiles
+    makeWrapper
+    pkg-config
     scons
-    udev
+  ];
+
+  buildInputs = [
+    alsa-lib
+    freetype
+    libGLU
+    libpulseaudio
     libX11
     libXcursor
+    libXext
+    libXfixes
+    libXi
     libXinerama
     libXrandr
     libXrender
-    libXi
-    libXext
-    libXfixes
-    freetype
     openssl
-    alsa-lib
-    libpulseaudio
-    libGLU
-    zlib
+    udev
     yasm
+    zlib
   ];
 
-  patches = [ ./pkg_config_additions.patch ./dont_clobber_environment.patch ];
+  shouldAddLinkFlagsToPulse = true;
+
+  patches = map (rp: ./patches + rp) [
+    # The version of SConstruct in the godot source appends the OS's PATH to the Scons PATH,
+    # but because it is an append, the Scons PATH takes precedence.  The Scons PATH contains a
+    # bunch of standard Linux paths like /usr/bin, so if they happen to contain versions of any
+    # build-time dependencies of Godot, they will be used instead of the Nix version of them.
+    #
+    # This patch simply replaces the entire Scons environment (including the PATH) with that
+    # of the OS. This isn't as surgical as just fixing the PATH, but it seems to work, and
+    # seems to be the Nix community's current strategy when using Scons.
+    /SConstruct/dontClobberEnvironment.patch
+  ];
 
   enableParallelBuilding = true;
+  buildPlatform = "x11";
+  shouldBuildTools = true;
+  buildTarget = "release_debug";
 
-  sconsFlags = [ "target=release_debug" "platform=x11" ];
-  preConfigure = ''
-    sconsFlags+=" ${
-      lib.concatStringsSep " "
-      (lib.mapAttrsToList (k: v: "${k}=${builtins.toJSON v}") options)
-    }"
-  '';
+  shouldUseLinkTimeOptimization = self.buildTarget == "release";
+
+  sconsFlags = [
+    "platform=${self.buildPlatform}"
```

```suggestion
  sconsFlags = let
      "i686-linux" = "x86_32";
      "x86_64-linux" = "x86_64";
      "aarch64-linux" = "arm64";
    }.${stdenv.hostPlatform.system};
  in [
    "arch=${godotArch}"
    "platform=${self.buildPlatform}"
```

This seems to fix aarch64-linux build as godot expects a custom `arch=` to be passed in (well it has an "auto" value by default, but I guess it's detecting something wrong?)

---

I'll make this change, and test to make sure it didn't _break_ anything with my x86_64 machine.  However, I don't have a way of testing that it will actually work on an aarch64 machine.

---

Let me know and I can test! I just added another change needed below and then aarch64 at least builds (I didn't test that anything functions)